solver/src/naive_solver/multi_order_solver.rs

```diff
@@ -0,0 +1,369 @@
+    Ok(SinglePairSettlement {
+        clearing_prices: maplit::hashmap! {
+            shortage.address => uniswap_in,
+            excess.address => uniswap_out,
+        },
+        trades: orders.into_iter().map(Trade::fully_matched).collect(),
+        interaction,
+    })
+}
+
+fn split_into_contexts(
+    orders: impl Iterator<Item = OrderCreation>,
+    reserves: &HashMap<Address, U256>,
+) -> Result<(TokenContext, TokenContext)> {
+    let mut contexts = HashMap::new();
+    for order in orders {
+        contexts.entry(order.buy_token).or_insert(TokenContext {
+            address: order.buy_token,
+            reserve: u256_to_bigint(
+                reserves
+                    .get(&order.buy_token)
+                    .ok_or_else(|| anyhow!("No reserve for token {}", &order.buy_token))?,
+            ),
+            ..Default::default()
+        });
+        contexts.entry(order.sell_token).or_insert(TokenContext {
+            address: order.sell_token,
+            reserve: u256_to_bigint(
+                reserves
+                    .get(&order.sell_token)
+                    .ok_or_else(|| anyhow!("No reserve for token {}", &order.sell_token))?,
+            ),
+            ..Default::default()
+        });
+        match order.kind {
+            OrderKind::Buy => {
+                contexts.get_mut(&order.buy_token).unwrap().buy_volume +=
+                    u256_to_bigint(&order.buy_amount)
+            }
+            OrderKind::Sell => {
+                contexts.get_mut(&order.sell_token).unwrap().sell_volume +=
+                    u256_to_bigint(&order.sell_amount)
+            }
+        }
+    }
+    if contexts.len() != 2 {
+        return Err(anyhow!("Orders contain more than two tokens"));
+    }
+    let mut contexts = contexts.drain().map(|(_, v)| v);
+    Ok((contexts.next().unwrap(), contexts.next().unwrap()))
+}
+
+/**
+ * Given information about the shortage token (the one we need to take from Uniswap) and the excess token (the one we gie to Uniswap), this function
+ * computes the exact out_amount required from Uniswap to perfectly match demand and supply at the effective Uniswap price (the one used for that in/out swap).
+ *
+ * The derivation of this formula is described in https://docs.google.com/document/d/1jS22wxbCqo88fGsqEMZgRQgiAcHlPqxoMw3CJTHst6c/edit
+ * It assumes GP fee (φ) to be 1 and Uniswap fee (Φ) to be 0.997
+ */
+fn compute_uniswap_out(shortage: &TokenContext, excess: &TokenContext) -> BigInt {
+    let numerator_minuend = 997 * (&excess.sell_volume - &excess.buy_volume) * &shortage.reserve;
+    let numerator_subtrahend =
+        1000 * (&shortage.sell_volume - &shortage.buy_volume) * &excess.reserve;
+    let denominator = (1000 * &excess.reserve) + (997 * (&excess.sell_volume - &excess.buy_volume));
+
+    (numerator_minuend - numerator_subtrahend) / denominator
```

Comment:
The two BigInt divisions (here and in compute_uniswap_in) will round down. Just pointing that out in case it isn't intentional as I don't fully understand the algorithm from the doc yet so it's not clear to me if that is intended.
There is also a BigRational type if that's helpful.

Reply:
We need to get a full number of out tokens for the uniswap payload. My plan was to deal with rounding errors by introducing a very small arbitrary fee (oba fee in the doc). This way we should be able to retain some small surplus in the contract and defend a bit against race conditions. If we simply round up here, my worry is that the in amount will also be larger and we might not have enough sell amounts on the other side of the trade.